DietPi-Software | Enhancement Fail2Ban

Joulinar opened the pull request:
At the moment Fail2Ban is failing after fresh installation with following error message

```
Oct 07 17:46:19 DietPiVM1 fail2ban-server[412]:  Failed during configuration: Bad value substitution: option 'filter' in section 'dropbear' contains an interpolation key 'mode' which is not a valid option name. Raw value: '%(__name__)s[mode=%(mode)s]'
```

Same behaviour on RPI3B+ as well as VM system. It is needed to add `mode = normal` before `filter` to get it working.

**Status**:  Ready 
- [x] Add Enhancement 
- [x] tested on VM 
- [x] tested on RPi 

**Reference**: https://dietpi.com/phpbb/viewtopic.php?f=11&t=8147

**Commit list/description**:
- DietPi-Software | Enhancement Fail2Ban as it is failing after fresh installation

Joulinar committed:
DietPi-Software | Enhancement Fail2Ban

atm Fail2Ban is failing after fresh installation

MichaIng commented:
Many thanks. Jep that matches the default package configs as well. Strange that this was never recognised before. Did it really lead to a failing service in your case? The user who reported this has still failing state 🤔.

Joulinar commented:
No it's working for the user now. That was his last statement

MichaIng commented:
Okay, maybe the posted status was before applying the fix.

When this leads to a failing fail2ban we should fix it during update as well, what you think?
```
[[ -f '/etc/fail2ban/jail.conf' ]] && GCI_PRESERVE=1 G_CONFIG_INJECT 'mode[[:blank:]]*=' 'mode = normal' /etc/fail2ban/jail.conf '\[DEFAULT\]'
```
This only adds the setting, if it not yet present, right below the `[DEFAULT]` section definition. Any manually added config will be preserved.

Joulinar commented:
ok good idea to have it implemented during DietPi update. What is strange for me, that we don't have more people reporting it. Maybe something that was introduced during one of the last Fail2Ban release?

MichaIng commented:
I'm also wondering. On Buster there has not been a release lately. However, regardless if it leads to crashes or not, it makes sense to define a variable that is used, even just with a dummy value, as well to expose it for users who want to check what can be configured and which plugin can use it in which way. Actually I was thinking to revert to the package default config, which has a bunch of nice explanations. I think it enables a few filters which are rarely required, but we can disable those in case to reduce CPU ran RAM usage (to scrape the logs and hold the database), at best based on installs.

MichaIng committed:
v6.33

+ DietPi-Patch | Fix potentially failing Fail2Ban service due to missing variable declaration

MichaIng committed:
v6.33

+ CHANGELOG | Fail2Ban: Resolved an issue where the service could have failed to start due to a missing variable declaration in our default config

Joulinar commented:
something to test how memory consumption is developing if we would use default config

MichaIng commented:
Jep, however not for this version anymore 😄.

cocoflan commented:
Same here just tested the install get this error:
Ather installing and rebooting the system:

when running command fail2ban-client status
 Failed to access socket path: /var/run/fail2ban/fail2ban.sock. Is fail2ban running?

MichaIng commented:
Okay, so definitely an important fix then!

cocoflan commented:
Is it not better to use tcpwrapper for this type of security?

cocoflan commented:
I installed fail2ban from the apk packages and it is working ok!

Joulinar commented:
Well that's a different story and not related to the missing variable in Fail2ban 😉

cocoflan commented:
Understanding Fail2Ban configuration file

There are two main configuration files in Fail2Ban: /etc/fail2ban/fail2ban.conf and /etc/fail2ban/jail.conf. Let me explain what they do.

/etc/fail2ban/fail2ban.conf: This is the configuration file for the operational settings of the Fail2Ban daemon. Settings like loglevel, log file, socket and pid file is defined here.

/etc/fail2ban/jail.conf: This is where all the magic happens. This is the file where you can configure things like default ban time, number of reties before banning an IP, whitelisting IPs, mail sending information etc. Basically you control the behavior of Fail2Ban from this file.

Now before you go and change these files, Fail2Ban advise to make a copy with .local file for these conf files. It’s because the default conf files can be overwritten in updates and you’ll lose all your settings.

sudo cp /etc/fail2ban/fail2ban.conf /etc/fail2ban/fail2ban.local
sudo cp /etc/fail2ban/jail.conf /etc/fail2ban/jail.local

MichaIng commented:
> It’s because the default conf files can be overwritten in updates and you’ll lose all your settings.

Those files are never overwritten by APT, as all files inside `/etc/fail2ban` are treated as "conffiles", where edited version are always left in place by default. Using `.local` is hence not to preserve customisations but to allow APT upgrades to update the defaults.

If we evaluate the defaults from the package to be mostly fine (should be) and only need to touch a small amount of these (if any), I'd go with `/etc/fail2ban/fail2ban.d/97_dietpi.conf` and `/etc/fail2ban/jail.d/97_dietpi.conf` (this kind of prefixing taken from man page, no idea what most common way is in the real world?) to apply DietPi customisation and leave the `.local` files (aside of the still available `.d/` for snippets) for the admin. On specific software installs we could add additional configs like `/etc/fail2ban/jail.d/98_dietpi-samba.conf` to automatically enable and in case configure additional jails. This would follow what I was thinking to implement on a regular basis:
- Use `*.d/` for all DietPi config overrides, where possible, prefixing all files with `dietpi`.
- If priority prefixes are generally used, apply `97` for general DietPi overrides, to override most package configs but leave space for admin.
- If a software install or config choice implies another config/override, use `98` to override the DietPi base override and append the software or config file name.
- So `99` is still available for local admin to override everything (aside of just skipping the integers and start with [a-z] which always has higher priority due to alphanumerical order 😉).

Joulinar commented:
@MichaIng 
I'm not sure if this is working well
```
[[ -f '/etc/fail2ban/jail.conf' ]] && GCI_PRESERVE=1 G_CONFIG_INJECT 'mode[[:blank:]]*=' 'mode = normal' /etc/fail2ban/jail.conf '\[DEFAULT\]'
```

It seems it's going to place the `mode = normal` at the end, because there is a hashed `#mode = normal` existing within `[sshd]` section. And this is just removing the `#`. But the line would need to be placed before `filter`. It's working only, if there is no `mode = normal` present at all.

MichaIng commented:
You are totally right, I only tested on our server where this commended mode below `[sshd]` is net yet present 😅.
**EDIT: Okay this does it: https://github.com/MichaIng/DietPi/commit/a19e329c72dabd8582fc5c43c31376fea5d3dc9a**

Joulinar commented:
tested on `dev` branch and it seems to be working.
